Integrate LFQA with Haystack - inferencing

vblagoje opened the pull request:
**Proposed changes**:

Implements [LFQA](https://github.com/deepset-ai/haystack/issues/914) inferencing based on Yacine Jernite's [work](https://yjernite.github.io/lfqa.html) and two models from HF hub: https://huggingface.co/yjernite/retribert-base-uncased and https://huggingface.co/yjernite/bart_eli5

The implementation consists of a new Retriever and a new Generator. 

**Retriever**

Retribert [model](https://huggingface.co/transformers/v4.4.2/model_doc/retribert.html) is used for matching similar query/passage documents. As Retribert model projects queries and passages into 128 dimensions the document store used in retriever tests had to be adjusted to use 128 dimensions. Note that the model projection into 128 dimensions is not configurable. Of course, one can bake a new Retribert model with Haystacks default dimension of 768 and perhaps that task could be added in the next stage when we add retriever training. 


**Generator**

A new generator is a classic seq2seq generator for auto-regressive language models. These models rely on [GenerationMixin's](https://huggingface.co/transformers/main_classes/model.html?highlight=generationmixin#transformers.generation_utils.GenerationMixin) generate method to generate text. Any models tagged with text2text-generation [tag](https://huggingface.co/models?pipeline_tag=text2text-generation) could potentially be used in such a generator. I chose this approach because the existing RAGenerator relies on HF classes that are not GenerationMixin subclasses but use their own RagSequenceForGeneration and RagTokenForGeneration class structure that indeed have a similar but different generate method signatures. To accommodate BartELI5 generation specifics a Seq2SeqGenerator abstract class has been extended with BartEli5Generator which prepares and encodes inputs specifically for https://huggingface.co/yjernite/bart_eli5 model.

As per suggestion by @tholor and @lalitpagaria I incorporated both Retriever and Generator into existing design. The new retriever has been integrated into EmbeddingRetriever. As EmbeddingRetriever hides a few different underlying approaches 
to  query and passage encoding, adding yet another Retribert encoder triggered a bit of a code smell that I remedied with slight refactoring that isolates components a bit better and accommodates potential new encoding approaches in the future. 

Adding unit tests proved to be a bit tricky due to 128 dimension projection requirement. I used the [approach](https://docs.pytest.org/en/latest/how-to/fixtures.html#using-markers-to-pass-data-to-fixtures) of using markers to pass data to fixtures. If reviewers have a suggestion how to accomplish this task better I would love to know. Looking forward to your feedback. 

**Status (please check what you already did)**:
- [x] First draft (up for discussions & feedback)
- [x] Added tests
- [x] Final code
- [ ] Updated documentation

tholor commented on haystack/generator/seq2seq.py:
```diff
@@ -0,0 +1,204 @@
+    def __init__(
+            self,
+            model_name_or_path: str,
+            retriever: Optional[BaseRetriever] = None,
```
How would we use a retriever here? From my understanding we just condition our generation on texts (and not the embeddings from an external retriever as in RAG)

vblagoje replied:
Hey @tholor 

Thanks for the praise. Most likely it is a fresh set of eyes looking at the problem :-)

> Let's tackle this before diving into a low-level review.

I agree with you. Let's find the best broad design strokes before diving into details.

> Is that also your understanding or do I miss more specialties here? Having clarity here will help us to settle on the design.

You are right on the spot. The key method interfaces seem to be embedding queries/passages. As Yacine's Retribert also distinguishes encoding queries and passage into different projects (using different methods as well) it seemed natural to isolate these high-level signatures. 

>  Having one class per model architecture could easily end up in a mess. That's why we don't have a RobertaReader, BERTReader etc.

I agree here except that I don't not foresee explosion of new EmbeddingRetriever subclasses/implementations as the current default implementation covers a lot of LM use cases, SentenceTransformer covers its use case, and finally Retribert its own. I am open to new suggestions here, let's find the best design approach. 

>A bit related to the above design questions: Is there a meaningful way to avoid the very model and dataset specific BartEli5Generator? I I think it will become eventually very messy to have one class per model. I can see that the only custom part here is the preparation of the input documents with < P>. 

You are right, very likely it will, but I am not sure how to work around it? I foresee the model input preparation to be the only model specific part of the code - think T5 task directives, BART placeholders etc etc, while everything else stays the same (the generate method invocation and parameters from GenerationMixin)

> We probably want to move Seq2seq to generators/transformers.py as the modules are currently rather divided by framework than by architecture and the RAG implementation is also in generators/transformers.py

No problem, will do, good decision.

vblagoje replied:
Yes, no need, a total oversight on my behalf here.

tholor reviewed (commented):
Hey @vblagoje , 

I had a first look at your PR and only focussed on the big design decisions so far. Here are some thoughts / open questions that hopefully help us to move forward on the general design. Let's tackle this before diving into a low-level review. 

Retriever: 
- I think your approach of using AutoModel + AutoTokenizer + InferenceProcessor is smart and could even work well for other models that we don't acticely support yet in FARM :)
- I agree that the current usage of EmbeddingRetriever with two different frameworks (FARM + sentence transformers) is already a bit ugly and it becomes worse with adding "retribert" to the mix as we mix frameworks with model architectures then
- However, I am still wondering what the best abstraction for this retriever is. Having one class per model architecture could easily end up in a mess. That's why we don't have a RobertaReader, BERTReader etc. 
Let me try to share my current understanding. From a quick look at the code, the special thing about retribert seems that it can have one or two BERT encoders, and even in the case of a single encoder there are two different projection layers on the very top of the network (one for queries, one for doc).
To distinguish which projection layer to use, they added in transformers separate `embed_answers()` and `embed_queries()` methods for retribert. We want to use these methods, but this kinda makes it a "special case" not handled by FARM's standard inference and not even handled in HF transformers' pipelines, I believe. That's why you created a dedicated class. 
Is that also your understanding or do I miss more specialties here? Having clarity here will help us to settle on the design.


Generator
- Great to have a more general seq2seq generator class!
- A bit related to the above design questions: Is there a meaningful way to avoid the very model and dataset specific `BartEli5Generator`? I think it will become eventually very messy to have one class per model. I can see that the only custom part here is the preparation of the input documents with `< P>`. How would this method look like for other seq2seq models from HF transformers? Can we infer the special token from the model / tokenizer config or even use a helper method from HF transformers? 
- We probably want to move Seq2seq to `generators/transformers.py` as the modules are currently rather divided by framework than by architecture and the RAG implementation is also in `generators/transformers.py` 

Happy to dive into the discussion and tackle the open questions one after the other!

vblagoje commented:
Author
Hey @tholor

Thanks for the praise. Most likely it is a fresh set of eyes looking at the problem :-)

> Let's tackle this before diving into a low-level review.

I agree with you. Let's find the best broad design strokes before diving into details.

> Is that also your understanding or do I miss more specialties here? Having clarity here will help us to settle on the design.

You are right on the spot. The key method interfaces seem to be embedding queries/passages. As Yacine's Retribert also distinguishes encoding queries and passage into different projects (using different methods as well) it seemed natural to isolate these high-level signatures.

> Having one class per model architecture could easily end up in a mess. That's why we don't have a RobertaReader, BERTReader etc.

I agree here except that I don't not foresee explosion of new EmbeddingRetriever subclasses/implementations as the current default implementation covers a lot of LM use cases, SentenceTransformer covers its use case, and finally Retribert its own. I am open to new suggestions here, let's find the best design approach.

> A bit related to the above design questions: Is there a meaningful way to avoid the very model and dataset specific BartEli5Generator? I I think it will become eventually very messy to have one class per model. I can see that the only custom part here is the preparation of the input documents with < P>.

You are right, very likely it will, but I am not sure how to work around it? I foresee the model input preparation to be the only model specific part of the code - think T5 task directives, BART placeholders etc etc, while everything else stays the same (the generate method invocation and parameters from GenerationMixin)

> We probably want to move Seq2seq to generators/transformers.py as the modules are currently rather divided by framework than by architecture and the RAG implementation is also in generators/transformers.py

No problem, will do, good decision.

lalitpagaria commented on haystack/generator/seq2seq.py:
```diff
@@ -0,0 +1,204 @@
+logger = logging.getLogger(__name__)
+
+
+class Seq2SeqGenerator(BaseGenerator):
```
I think it is better to move it to generator/transformers.py

lalitpagaria commented on haystack/generator/seq2seq.py:
```diff
@@ -0,0 +1,204 @@
+        return {"query": query, "answers": generated_answers}
+
+
+class BartEli5Generator(Seq2SeqGenerator):
```
Not sure it is good idea to create separate class for `prepare_model_input` function.
I don't have much knowledge about model family though. @tholor your comment please

lalitpagaria commented on haystack/retriever/dense.py:
```diff
@@ -461,6 +462,7 @@ def __init__(
         pooling_strategy: str = "reduce_mean",
         emb_extraction_layer: int = -1,
         top_k: int = 10,
+        progress_bar: bool = True
```
Please doc string param

lalitpagaria commented on haystack/retriever/dense.py:
```diff
@@ -552,54 +520,208 @@ def retrieve(self, query: str, filters: dict = None, top_k: Optional[int] = None
+
+class _EmbeddingEncoder:
 
     def embed_queries(self, texts: List[str]) -> List[np.ndarray]:
```
Please make it `abstractfunction`

lalitpagaria commented on haystack/retriever/dense.py:
```diff
@@ -552,54 +520,208 @@ def retrieve(self, query: str, filters: dict = None, top_k: Optional[int] = None
-        return self.embed(texts)
+        pass
 
-    def embed_passages(self, docs: List[Document]) -> Union[List[str], List[List[str]]]:
```
Please make it `abstractfunction`

lalitpagaria commented on haystack/retriever/dense.py:
```diff
@@ -552,54 +520,208 @@ def retrieve(self, query: str, filters: dict = None, top_k: Optional[int] = None
+        pass
+
+
+class _EmbeddingEncoderFactory:
```
I liked this abstraction :)

vblagoje commented:
@tholor @lalitpagaria do you have any more comments or general feedback? Regarding documentation - what should I add?

tholor commented:
Oh okay, I thought you were still in the process of coding the next iteration to discuss. So, good that you pinged me :+1:  

- I am fine with the RetribertEncoder approach. I think separating out the Encoders from the EmbeddingRetriever makes sense and the Retribert is different enough in the model architecture to justify an own Encoder class. Will provide a more granular review here soon.
- I am still not super happy with the Generator, where we will have subclasses per model just to implement `prepare_model_input`. From my current understanding, we would probably soon support different model architectures (e.g. T5) and depending on their fine-tuning style different prefixes and document separators. However, the other solutions that come into my mind right now are not really nice either. Let me digest this a bit and I will come back to you with some alternatives that we can discuss.


Regarding documentation: 
- Docstrings seem almost complete already :+1: - Can you just add a few lines to the different encoder classes (RetribertEncoder etc.) explaining their purpose?
- We should extend the current usage page for the "Generator": https://haystack.deepset.ai/docs/latest/generatormd
- Ideally, we have a small Tutorial (similar to the notebook you already shared) - maybe using the Eli5 dataset. Alternatively we could also add it to the existing[ RAG tutorial](https://github.com/deepset-ai/haystack/blob/master/tutorials/Tutorial7_RAG_Generator.ipynb) and make it a big "Generative QA Tutorial" contrasting the two different approaches.

tholor commented:
Ok, as promised, I thought again about a reasonable way forward for the `BartEli5Generator` abstraction and also had a quick discussion with @oryx1729 about it. 

Here is a quick summary of our thoughts and our suggestion for the way forward: 
-  While being a bit cleaner from SE perspective the current inheritance is inconsistent with the other "building blocks" in Haystack where a model instance is usually configured on the parameter level (e.g. `model_name_or_path` for the reader) and not on the class level. This inconsistency could confuse users and makes documentation harder.
- While in python this seems a relatively small difference, it is a bigger one if you think about pipelines represenations in yml,  drawn as a graph or in a potential "pipeline designer" where you really pick nodes in a UI, stick them together to a UI and configure their parameters in a "detail view"
- We suggest therefore to not subclass the `Seq2SeqGenerator` and let the user specify the model of his/her choice via the arg `model_name_or_path`
- For `prepare_model_inputs()` we see two implementation options: a) we start with simple if-else blocks depending on the provided model string or class type of the model and worry about factoring it out if this list grows beyond 3-4 cases OR b) we directly design some private class like a `BartEli5InputConverter` that does the custom preparation and is initialized depending on the provided model string, class or optionally supplied by a user argument.
- With this the user interface would remain consisten to the other classes, while under the hood we could still factor some parts out later
- One other alternative that we considered: breaking down the current preparation step to some common parameters (e.g. `prefix`, `doc_sep_token` etc) that kinda describe the preprocessing independently from the actual model architecture. However, with this we still have the problem of matching "model name" with the appropriate parameters. Ideally, this should become part of a model_config in HF transformers so that this is really something closely tied to the model and easily read in a generic way. 

@vblagoje  Does this make sense to you?

vblagoje commented:
@tholor @oryx1729 your suggestions sound good. I agree subclass per use case is a design smell. I don't mind if-else use case approach and then wait for the community to provide different generation cases that we can address at that time. I would also not forget the converter class approach either as you hint. How about the __call__ interface converter, something like this:

```
class BartEli5Converter:
	def __init__(self):
		# some init here

	# Defining __call__ method
	def __call__(self, query: str, documents: List[Document], top_k: Optional[int] = None) -> Dict:
		# do conversion here
                conditioned_doc = "<P> " + " <P> ".join([d.text for d in documents])       
                query_and_docs = "question: {} context: {}".format(query, conditioned_doc)        
                return self.tokenizer([(query_and_docs, "A")], truncation=True, padding=True, return_tensors="pt").to(self.device)

# Create Seq2SeqGenerator
generator = Seq2SeqGenerator(model_name_or_path="yernite/bart_eli5", input_converter=BartEli5Converter())

```

In Seq2SeqGenerator impl we'd then invoke converter with the predefined parameters. I mean, it doesn't have to be __call__ approach at all, it could be a named abstract class but then again - no need to waste a class. Users can implement it almost inline for their specific use case. Either way, you guys let me know what you'd prefer, what is the most consistent and we'll go for that.

tholor commented:
Ok, great. Then let's go ahead in this direction! 
I like the idea of the `BartEli5Converter` with a __call__ method. I think the `input_converter` arg is also reasonable. Just to make sure that we are on the same page here: this arg should not be mandatory IMO as the user might not know which `model_name_or_path`  to pair with which converter and there's no reason to create this "mandatory obstacle" for users to read the documentation / docstring first to find out. So let's here still have some kind of "automatic if-else matching" that works on model string or model class level, determines the appropriate Converter and initializes it. We could trigger an exception if it is an unknown `model_name_or_path` where our matching logic fails and point the user to the option to supply a custom `InputConverter`.

tholor reviewed (approved):
Looking good to me! From my perspective, this is ready-to-merge once mypy is satisfied and CI is passing. @vblagoje could you please fix the typing issue there with the Callable? You can test this locally via ` mypy haystack --ignore-missing-imports`. Thx

vblagoje commented:
Should be fixed now @tholor Why don't we add some pre-commit scripts developers can run locally to avoid these issues in the future? I've also seen projects use https://pre-commit.com/ successfully.

vblagoje commented:
@lalitpagaria please focus on docstrings review and I will update the Generator page in the meantime. I'll share the notebook in the current form (without ELI5 dataset). HF datasets should soon support features where we can take a small subset of wiki snippets to demo the notebook similar to https://yjernite.github.io/lfqa.html but until then the demo is a bit cumbersome for general public IMHO. 

>Regarding documentation:

>Docstrings seem almost complete already 👍 - Can you just add a few lines to the different encoder classes > (RetribertEncoder etc.) explaining their purpose?
> We should extend the current usage page for the "Generator": https://haystack.deepset.ai/docs/latest/generatormd
> Ideally, we have a small Tutorial (similar to the notebook you already shared) - maybe using the Eli5 dataset. Alternatively 
> we could also add it to the existing RAG tutorial and make it a big "Generative QA Tutorial" contrasting the two different 
> approaches.
>

lalitpagaria commented on haystack/generator/transformers.py:
```diff
@@ -291,3 +292,181 @@ def predict(self, query: str, documents: List[Document], top_k: Optional[int] =
+
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
+        self.model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path).to(self.device)
+        self.model.eval()
```
just curious why this step is needed?

vblagoje replied:
The eval switch? Yeah, it's not, perhaps just to be explicit that the model is indeed in eval mode.

lalitpagaria commented on haystack/generator/transformers.py:
```diff
@@ -291,3 +292,181 @@ def predict(self, query: str, documents: List[Document], top_k: Optional[int] =
+    def __init__(
+            self,
+            model_name_or_path: str,
+            input_converter: Optional[Callable] = None,
```
@tholor how user will supply these param if creating pipeline via yaml?

vblagoje replied:
Yeah, not sure how that works exactly. LMK

tholor replied:
@oryx1729 You had a similar case recently with the retrievers where we supply a documentstore object. What is the requirement here for saving/loading yamls? Does it have to inherit from `BaseComponent`? 

I think in any case this could only work for now with predefined callables in the haystack namespace (like  `_BartEli5Converter`). We can tackle custom components later, maybe together with custom nodes in https://github.com/deepset-ai/haystack/issues/955

lalitpagaria commented on haystack/generator/transformers.py:
```diff
@@ -291,3 +292,181 @@ def predict(self, query: str, documents: List[Document], top_k: Optional[int] =
+            raise KeyError(f"Seq2SeqGenerator doesn't have input converter registered for {self.model_name_or_path}. "
+                           f"Provide custom converter for {self.model_name_or_path} in Seq2SeqGenerator initialization")
+
+        query_and_docs_encoded = converter(self.tokenizer, query, documents, top_k).to(self.device)
```
How to make sure user supplied Callable follow this syntax?
I understand Callable is very elegant but these is chance of error.
How about having Base class which have abstract function with this signature?

vblagoje replied:
When you say "chance of error" are you referring to a user potentially mismatching the signature of Callable? I can see that point but we'd like to keep this Callable relatively hidden and for super-user customization. Your call @tholor

tholor replied:
I would prefer going forward with the light-weight Callable. However, we should add the expected signature to the docstring of  the `Seq2SeqGenerator.__init__()` and maybe link `_BartEli5Converter` as an example.

To make it a bit more failsafe, we could change the call to `converter(tokenizer=self.tokenizer, query=query, documents=documents, top_k=top_k).to(self.device)` and catch the `TypeError` with a descriptive error message if the converter doesn't support this signature :thinking:

vblagoje replied:
Great suggestion @tholor thank you

lalitpagaria commented on haystack/retriever/dense.py:
```diff
@@ -552,54 +522,210 @@ def retrieve(self, query: str, filters: dict = None, top_k: Optional[int] = None
+                              "For details see https://github.com/UKPLab/sentence-transformers ")
+        # pretrained embedding models coming from: https://github.com/UKPLab/sentence-transformers#pretrained-models
+        # e.g. 'roberta-base-nli-stsb-mean-tokens'
+        if retriever.use_gpu:
```
Can we make it util function it is used at multiple places. Some places we use pytorch conda check as well.

vblagoje replied:
Sounds good but that's suitable for another PR, right @lalitpagaria ?

lalitpagaria replied:
Yes agree

lalitpagaria commented on haystack/retriever/dense.py:
```diff
@@ -552,54 +522,210 @@ def retrieve(self, query: str, filters: dict = None, top_k: Optional[int] = None
-            emb = [r for r in emb]
-        return emb
 
+class _EmbeddingEncoder:
```
I am in favour to move all `EmbeddingEncoder` related code to another file. This file is becoming very big also very hard to navigate around code.

vblagoje replied:
I don't mind tbh, except IIRC we wanted to keep everything in dense.py including EmbeddingRetriever. Perhaps we can leave this for a future refactor @tholor ?

tholor replied:
Yes, let's leave this for a future refactor. I can see that we need to rethink the module structure here in a more holistic way.

lalitpagaria reviewed (changes requested):
Have few comments otherwise looks fine.

vblagoje commented:
@lalitpagaria I also added LFQA tutorials. Would you please check it as well? Not sure how exactly Haystack documentation gets generated so I could use some help here. @tholor I don't mind if you want to wait for a more impressive LFQA (with wikipages) similar the original Yacine's demo. In that case we can drop the tutorial for now. Or we can include the current one and improve it in the next iteration. Either way, LMK your decision.

tholor commented:
@vblagoje I left a few comments above. Can you please also resolve the conflicts with master branch? 


> I also added LFQA tutorials. Would you please check it as well? Not sure how exactly Haystack documentation gets generated so I could use some help here. @tholor I don't mind if you want to wait for a more impressive LFQA (with wikipages) similar the original Yacine's demo. In that case we can drop the tutorial for now. Or we can include the current one and improve it in the next iteration. Either way, LMK your decision.

@brandenchan can you please guide @vblagoje on what docs we want for LFQA and how to generate them? Please also have a look at the tutorial and decide if we want to merge it or first improve it in a separate PR.

vblagoje commented:
Rebased to master, fixed Callable type error reporting, added unit test, and fixed documentation in ce3a9b4 Let me know how we can prepare the documentation @brandenchan and this one should be good to go.

lalitpagaria reviewed (approved):
LGTM

brandenchan commented:
Hi I had a look at the documentation + tutorials side of things and it all looks good to me! I just did a little reformatting of comments and also regenerated some md files. Giving this the thumbs up for merge.

vblagoje commented:
@brandenchan Thanks a lot. Whoever has the privileges, please feel free squash the commits, format the final commit message, and execute the merge as you see fit.

tholor commented:
Awesome! Happy to see this one being merged. Thanks a lot for your work @vblagoje - I think it's a really solid piece now.
